main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -28,301 +28,346 @@ import ca.uwaterloo.flix.util.Validation._
 
     result.toSuccess
   }
 
+  /**
+   * Performs expression inlining on the given definition `def0`.
+   * Converts definition from OccurrenceAst to LiftedAst.
+   */
+  private def visitDef(def0: OccurrenceAst.Def)(implicit flix: Flix): LiftedAst.Def = {
+    val convertedExp = visitExp(def0.exp, Map.empty)
+    val fparams = def0.fparams.map {
+      case OccurrenceAst.FormalParam(sym, mod, tpe, loc) => LiftedAst.FormalParam(sym, mod, tpe, loc)
+    }
+    LiftedAst.Def(def0.ann, def0.mod, def0.sym, fparams, convertedExp, def0.tpe, def0.loc)
+  }
 
   /**
    * Performs inlining operations on the expression `exp0` of type OccurrenceAst.Expression.
    * Returns an expression of type Expression
    */
-  private def visitExp(exp0: OccurrenceAst.Expression, subst0: Map[Symbol.VarSym, Expression]): Expression = exp0 match {
-    case Expression.Unit(_) => exp0
+  private def visitExp(exp0: OccurrenceAst.Expression, subst0: Map[Symbol.VarSym, Expression]): LiftedAst.Expression = exp0 match {
+    case OccurrenceAst.Expression.Unit(loc) => LiftedAst.Expression.Unit(loc)
 
-    case Expression.Null(_, _) => exp0
+    case OccurrenceAst.Expression.Null(tpe, loc) => LiftedAst.Expression.Null(tpe, loc)
 
-    case Expression.True(_) => exp0
+    case OccurrenceAst.Expression.True(loc) => LiftedAst.Expression.True(loc)
 
-    case Expression.False(_) => exp0
+    case OccurrenceAst.Expression.False(loc) => LiftedAst.Expression.False(loc)
 
-    case Expression.Char(_, _) => exp0
+    case OccurrenceAst.Expression.Char(lit, loc) => LiftedAst.Expression.Char(lit, loc)
 
-    case Expression.Float32(_, _) => exp0
+    case OccurrenceAst.Expression.Float32(lit, loc) => LiftedAst.Expression.Float32(lit, loc)
 
-    case Expression.Float64(_, _) => exp0
+    case OccurrenceAst.Expression.Float64(lit, loc) => LiftedAst.Expression.Float64(lit, loc)
 
-    case Expression.Int8(_, _) => exp0
+    case OccurrenceAst.Expression.Int8(lit, loc) => LiftedAst.Expression.Int8(lit, loc)
 
-    case Expression.Int16(_, _) => exp0
+    case OccurrenceAst.Expression.Int16(lit, loc) => LiftedAst.Expression.Int16(lit, loc)
 
-    case Expression.Int32(_, _) => exp0
+    case OccurrenceAst.Expression.Int32(lit, loc) => LiftedAst.Expression.Int32(lit, loc)
 
-    case Expression.Int64(_, _) => exp0
+    case OccurrenceAst.Expression.Int64(lit, loc) => LiftedAst.Expression.Int64(lit, loc)
 
-    case Expression.BigInt(_, _) => exp0
+    case OccurrenceAst.Expression.BigInt(lit, loc) => LiftedAst.Expression.BigInt(lit, loc)
 
-    case Expression.Str(_, _) => exp0
+    case OccurrenceAst.Expression.Str(lit, loc) => LiftedAst.Expression.Str(lit, loc)
 
-    case Expression.Var(sym, _, _) => subst0.get(sym).fold(exp0)(visitExp(_, subst0))
+    case OccurrenceAst.Expression.Var(sym, tpe, loc) =>
+      val e = subst0.get(sym)
+      e.fold[LiftedAst.Expression](LiftedAst.Expression.Var(sym, tpe, loc)) {
```

Comment:
Using a fold here is "overkill". I think you should just do subst0.get(sym) and a pattern match. Then each case would be  easier to understand.